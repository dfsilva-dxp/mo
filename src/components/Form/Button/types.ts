import { ButtonHTMLAttributes, ReactNode } from "react";

export type ButtonProps = {
  children?: ReactNode;
  size?: "sm" | "md" | "lg";
  fullSize?: boolean;
  icon?: ReactNode;
  color?: "primary" | "darkgray" | "lightgray";
} & ButtonHTMLAttributes<HTMLButtonElement>;
