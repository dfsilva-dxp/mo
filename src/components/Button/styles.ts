import styled from "styled-components";

export const Button = styled.button`
  height: var(--px-48);
  padding: 0 2.25rem;
  vertical-align: middle;
  user-select: none;
  border-radius: var(--px-4);
  font-size: var(--ft-14);
  color: var(--gray-500);
  line-height: 1.5;
  font-weight: 500;
  text-transform: uppercase;
  letter-spacing: 1px;
  transition: background-color 0.15s ease-in-out, color 0.15s ease-in-out;

  &.primary {
    background-color: var(--yellow-500);
    border: 1px solid transparent;
    &:not([disabled]):hover {
      background-color: var(--yellow-400);
    }
  }

  &.secondary {
    border: 1px solid var(--gray-150);
    background-color: var(--gray-750);
    &:not([disabled]):hover {
      background-color: var(--gray-700);
    }
  }
`;
