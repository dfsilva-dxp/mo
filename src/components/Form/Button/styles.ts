import styled, { css, DefaultTheme } from "styled-components";
import { ButtonProps } from "./types";

type BtnProps = Pick<ButtonProps, "size" | "fullSize" | "color">;

const buttonModifiers = {
  sm: (theme: DefaultTheme) => css`
    padding: 0.4rem 1rem;
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};

    svg {
      font-size: ${theme.font.sizes.small};
    }
  `,
  md: (theme: DefaultTheme) => css`
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
    font-size: ${theme.font.sizes.small};

    svg {
      font-size: ${theme.font.sizes.medium};
    }
  `,
  lg: (theme: DefaultTheme) => css`
    padding: ${theme.spacings.xxsmall} ${theme.spacings.small};
    height: 4rem;
    font-size: ${theme.font.sizes.medium};

    svg {
      font-size: 2rem;
    }
  `,
  fullSize: () => css`
    width: 100%;
  `,
  primary: () => css`
    background: linear-gradient(180deg, #f2a516 0%, #ffc727 50%);

    &:hover {
      background: linear-gradient(180deg, #ffc727 0%, #f2a516 50%);
    }
  `,
  darkgray: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: linear-gradient(180deg, #1f2029 0%, #022b3a 50%);

    &:hover {
      background: linear-gradient(180deg, #022b3a 0%, #1f2029 50%);
    }
  `,
  lightgray: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: linear-gradient(180deg, #616480 0%, #797d9a 50%);

    &:hover {
      background: linear-gradient(180deg, #797d9a 0%, #616480 50%);
    }
  `
};

export const Btn = styled.button<BtnProps>`
  ${({ theme, size, fullSize, color }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    gap: ${theme.spacings.xxsmall};
    min-width: 64px;
    font-size: initial;
    font-weight: ${theme.font.semibold};
    color: ${theme.colors.gray750};
    border: 0;
    border-radius: ${theme.border.radius};

    ${!!size && buttonModifiers[size](theme)};
    ${fullSize && buttonModifiers.fullSize()};
    ${!!color && buttonModifiers[color](theme)};
  `}
`;
