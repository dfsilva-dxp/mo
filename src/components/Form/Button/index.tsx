import { ButtonProps } from "./types";

import { Btn } from "./styles";

export default function Button({
  children,
  size = "md",
  fullSize = false,
  color = "primary",
  icon,
  ...props
}: ButtonProps) {
  return (
    <Btn size={size} fullSize={fullSize} color={color} {...props}>
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </Btn>
  );
}
