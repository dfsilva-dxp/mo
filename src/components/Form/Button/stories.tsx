import { Story, Meta } from "@storybook/react/types-6-0";
import { ComponentProps } from "react";
import { TiShoppingCart } from "react-icons/ti";

import Button from ".";

export default {
  title: "Form/Button",
  component: Button,
  argTypes: {
    size: {
      options: ["sm", "md", "lg"],
      control: { type: "select" }
    },
    color: {
      options: ["primary", "darkgray", "lightgray"],
      control: { type: "select" }
    },
    icon: {
      control: { type: "" }
    }
  },
  args: {
    size: "md",
    color: "primary",
    fullSize: false
  }
} as Meta;

export const Basic: Story<ComponentProps<typeof Button>> = ({ ...props }) => (
  <Button {...props}>Enter</Button>
);

export const WithIcon: Story<ComponentProps<typeof Button>> = ({
  ...props
}) => <Button {...props}>Buy now</Button>;

WithIcon.args = {
  icon: <TiShoppingCart />
};
