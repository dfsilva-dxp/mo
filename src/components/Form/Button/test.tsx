import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TiShoppingCart } from "react-icons/ti";

import { renderWithTheme } from "utils/tests/helpers";

import Button from ".";

describe("Button Component", () => {
  it("should render a button", () => {
    const { container } = renderWithTheme(<Button>Button</Button>);

    expect(screen.getByRole("button", { name: /Button/i })).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the md button by default", () => {
    renderWithTheme(<Button>Button</Button>);

    expect(screen.getByRole("button", { name: /Button/i })).toHaveStyle({
      padding: "0.8rem 1.6rem",
      "font-size": "1.4rem"
    });
  });

  it("should render the small button when sm size is passed", () => {
    renderWithTheme(<Button size="sm">Button</Button>);

    expect(screen.getByRole("button", { name: /Button/i })).toHaveStyle({
      padding: "0.4rem 1rem",
      height: "3rem",
      "font-size": "1.2rem"
    });
  });

  it("should render the large button when lg size is passed", () => {
    renderWithTheme(<Button size="lg">Button</Button>);

    expect(screen.getByRole("button", { name: /Button/i })).toHaveStyle({
      padding: "0.8rem 2rem",
      height: "4rem",
      "font-size": "1.6rem"
    });
  });

  it("should render the full width button when full size is passed", () => {
    renderWithTheme(<Button fullSize>Button</Button>);

    expect(screen.getByRole("button", { name: /Button/i })).toHaveStyle({
      width: "100%"
    });
  });

  it("should render without icon", () => {
    renderWithTheme(<Button>Buy now</Button>);

    expect(screen.queryByTestId("icon")).not.toBeInTheDocument();
  });

  it("should render with icon", () => {
    renderWithTheme(
      <Button icon={<TiShoppingCart data-testid="icon" />}>Buy now</Button>
    );

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should is accessible by tab", () => {
    renderWithTheme(<Button>Button</Button>);

    const button = screen.getByRole("button", { name: /Button/i });

    expect(document.body).toHaveFocus();
    userEvent.tab();

    expect(button).toHaveFocus();
  });
});
