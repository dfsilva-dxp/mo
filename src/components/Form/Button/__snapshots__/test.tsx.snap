// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render a button 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 0.8rem;
  min-width: 64px;
  font-size: initial;
  font-weight: 600;
  color: #404152;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem 1.6rem;
  font-size: 1.4rem;
  background: linear-gradient(180deg,#f2a516 0%,#ffc727 50%);
}

.c0 svg {
  font-size: 1.6rem;
}

.c0:hover {
  background: linear-gradient(180deg,#ffc727 0%,#f2a516 50%);
}

<button
  class="c0"
  color="primary"
>
  <span>
    Button
  </span>
</button>
`;
